packages/react-instantsearch-core/src/widgets/DynamicWidgets.tsx

```diff
@@ -0,0 +1,52 @@
+import React, { Fragment, ReactChild, ReactNode } from 'react';
+import { getDisplayName } from '../core/utils';
+import connectDynamicWidgets from '../connectors/connectDynamicWidgets';
+
+function getAttribute(component: ReactChild): string | undefined {
+  if (typeof component !== 'object') {
+    return undefined;
+  }
+
+  if (component.props.attribute) {
+    return component.props.attribute;
+  }
+  if (Array.isArray(component.props.attributes)) {
+    return component.props.attributes[0];
+  }
+  if (component.props.children) {
+    return getAttribute(React.Children.only(component.props.children));
+  }
+
+  return undefined;
+}
+
+type DynamicWidgetsProps = {
+  children: ReactNode;
+  attributesToRender: string[];
+};
+
+function DynamicWidgets({ children, attributesToRender }: DynamicWidgetsProps) {
+  const widgets: Map<string, ReactChild> = new Map();
+
+  React.Children.forEach(children, child => {
+    const attribute = getAttribute(child);
+    if (!attribute) {
+      throw new Error(
+        `attribute could not be found for ${getDisplayName(child)}`
+      );
+    }
+    widgets.set(attribute, child);
+  });
+
+  // on initial render this will be empty, but React InstantSearch keeps
+  // search state for unmounted components in place, so routing works.
```

Comment:
can you point me to which part of the code enables this?

Reply:
React InstantSearch keeps searchState always as it is, unless cleanUp of a component has been called (when it unmounts). That means that whatever state is available over refinementList for example, it will not be removed until that component unmounts. Initially the componeent doesn't _need_ to be visible

Reply:
Thanks for clarification.